src/slide-image-source.cc

```diff
@@ -0,0 +1,159 @@
+
+// Scoped clean-up of c objects.
+struct ScopeGuard {
+    ScopeGuard(std::function<void()> f) : f_(f) {}
+    ~ScopeGuard() { f_(); }
+    std::function<void()> f_;
+};
+
+static inline bool invalid_dimensions(int64_t width, int64_t height) {
+    return (width < 0 || height < 0) ||
+        (width > std::numeric_limits<int>::max() ||
+         height > std::numeric_limits<int>::max());
+}
+
+bool SlideImageSource::LoadAndScale(const DisplayOptions &opts, int, int) {
+    options_ = opts;
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    int64_t width, height;
+    int target_width;
+    int target_height;
+    double scaling;
+
+    // First check if there's a thumbnail with enough resolution.
+    const char* const* associated_image_names = openslide_get_associated_image_names(osr);
+    char const *target_associated_image = NULL;
+    while (*associated_image_names) {
+        const char *name = *associated_image_names;
+        openslide_get_associated_image_dimensions(osr, name, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        if (!strcmp(name, "thumbnail")) {
+            // Compute the scaling
+            CalcScaleToFitDisplay(width, height, opts, false,
+                                  &target_width, &target_height);
+            scaling = (double) target_width / width;
+            target_associated_image = (scaling < 1.0) ? name : NULL;
+            break;
+        }
+        associated_image_names++;
+    }
+
+    // Generate the native image framebuffer
+    timg::Framebuffer decode_image(width, height);
+    if (target_associated_image)
+        openslide_read_associated_image(osr, target_associated_image,
+                                        (uint32_t *)decode_image.begin());
+    else {
+        // We need to use one of the layers of the pyramid.
+        // Get native the highest resolution image.
+        openslide_get_level0_dimensions(osr, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        const double scaling = (double) target_width / width;
+
+        // Upscaling? Unsure what to do.
+        if (scaling > 1.0)
+            return false;
+
+        // Get the best layer to downsample from.
+        const int32_t level = openslide_get_best_level_for_downsample(osr, scaling);
+        if (level < 0)
+            return false;
+
+        openslide_read_region(osr, (uint32_t *) decode_image.begin(), 0, 0, level, width, height);
+        const char *err = openslide_get_error(osr);
+        if (err)
+            return false;
+    }
+
+    // TODO: Unpremultiply
```

Comment:
did you confirm that the alpha thing is indeed premultiplied and this was not a confusion because `AlphaComposeBackground()` was not called ?

If so, I can give you a AlphaComposeBackground that does not do the multiplication.

Reply:
(Solved)

Reply:
so is this comment still needed now that things are figured out ?